File: Soulo/Models/JournalEntry.swift
import Foundation

struct JournalEntry: Identifiable, Codable, Sendable {
    let id: String
    let timestamp: TimeInterval
    let timezoneOffset: Int
    let durationMs: Int

    var audioFormat: String
    var audioEncrypted: Bool
    var audioFileSize: Int?

    var transcript: String?
    var transcriptStatus: ProcessingStatus
    var transcriptError: String?
    var transcriptMs: Int?

    var biomarkersJson: String?
    var biomarkersStatus: ProcessingStatus
    var biomarkersError: String?

    var emotionJson: String?
    var emotionStatus: ProcessingStatus
    var emotionError: String?

    var topicsJson: String?
    var topicsStatus: ProcessingStatus
    var topicsError: String?

    let appVersion: String
    let deviceModel: String?
    let osVersion: String?
    let createdAt: TimeInterval
    var updatedAt: TimeInterval

    init(
        id: String = UUID().uuidString,
        timestamp: TimeInterval = Date().timeIntervalSince1970,
        timezoneOffset: Int = TimeZone.current.secondsFromGMT() / 60,
        durationMs: Int,
        audioFormat: String = "aac",
        audioEncrypted: Bool = true
    ) {
        self.id = id
        self.timestamp = timestamp
        self.timezoneOffset = timezoneOffset
        self.durationMs = durationMs
        self.audioFormat = audioFormat
        self.audioEncrypted = audioEncrypted
        self.transcriptStatus = .pending
        self.biomarkersStatus = .pending
        self.emotionStatus = .pending
        self.topicsStatus = .pending
        self.appVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0"
        self.deviceModel = Self.deviceModel
        self.osVersion = UIDevice.current.systemVersion
        self.createdAt = Date().timeIntervalSince1970
        self.updatedAt = Date().timeIntervalSince1970
    }

    var biomarkers: VoiceBiomarkers? {
        get {
            guard let data = biomarkersJson?.data(using: .utf8) else { return nil }
            return try? JSONDecoder().decode(VoiceBiomarkers.self, from: data)
        }
        set {
            biomarkersJson = newValue.flatMap { try? String(data: JSONEncoder().encode($0), encoding: .utf8) }
        }
    }

    var emotion: EmotionalState? {
        get {
            guard let data = emotionJson?.data(using: .utf8) else { return nil }
            return try? JSONDecoder().decode(EmotionalState.self, from: data)
        }
        set {
            emotionJson = newValue.flatMap { try? String(data: JSONEncoder().encode($0), encoding: .utf8) }
        }
    }

    var topics: [TopicAnalysis]? {
        get {
            guard let data = topicsJson?.data(using: .utf8) else { return nil }
            return try? JSONDecoder().decode([TopicAnalysis].self, from: data)
        }
        set {
            topicsJson = newValue.flatMap { try? String(data: JSONEncoder().encode($0), encoding: .utf8) }
        }
    }

    private static var deviceModel: String? {
        var systemInfo = utsname()
        uname(&systemInfo)
        let mirror = Mirror(reflecting: systemInfo.machine)
        return mirror.children.compactMap { $0.value as? Int8 }.map { String(UnicodeScalar(UInt8($0))) }.joined()
            .trimmingCharacters(in: .controlCharacters)
    }
}

enum ProcessingStatus: String, Codable {
    case pending, processing, done, failed
}

File: android/app/src/main/java/com/soulo/app/models/JournalEntry.kt
package com.soulo.app.models

import kotlinx.serialization.Serializable

@Serializable
enum class ProcessingStatus { pending, processing, completed, failed }

@Serializable
data class JournalEntry(
    val id: String,
    val timestamp: Long,
    val durationMs: Long,
    val transcript: String? = null,
    val audioFile: String? = null,
    val isQuickEntry: Boolean = false,
    val biomarkers: VoiceBiomarkers? = null,
    val emotion: EmotionalState? = null,
    val topics: List<TopicAnalysis>? = null,
    val transcriptStatus: ProcessingStatus = ProcessingStatus.pending,
    val biomarkersStatus: ProcessingStatus = ProcessingStatus.pending,
    val emotionStatus: ProcessingStatus = ProcessingStatus.pending,
    val topicsStatus: ProcessingStatus = ProcessingStatus.pending,
    val appVersion: String = "1.0.0",
    val deviceModel: String = "",
    val osVersion: String = "",
    val createdAt: Long = System.currentTimeMillis() / 1000,
    val updatedAt: Long = System.currentTimeMillis() / 1000
) {
    val formattedDate: String
        get() = java.text.SimpleDateFormat("MMM d, yyyy", java.util.Locale.getDefault())
            .format(java.util.Date(timestamp * 1000))

    val durationFormatted: String
        get() {
            val secs = durationMs / 1000
            val m = secs / 60
            val s = secs % 60
            return "${m}m ${s}s"
        }

    companion object {
        val deviceModel: String get() = android.os.Build.MODEL
        val osVersion: String get() = android.os.Build.VERSION.RELEASE
    }
}
